### PR TITLE
Add Segment unsuppression API

### PR DESCRIPTION
**Pull request by bmedx**

### CODE REVIEW

File: scripts/bulk_unsuppress_segment_users.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+ERR_BAD_CONFIG = -2
+ERR_NO_CSV_FILE = -3
+ERR_UNSUPPRESSING_USERS = -4
+ERR_BAD_KEY = -5
+
+SCRIPT_SHORTNAME = 'bulk_unsuppress_segment_users'
+LOG = partial(_log, SCRIPT_SHORTNAME)
+FAIL = partial(_fail, SCRIPT_SHORTNAME)
+FAIL_EXCEPTION = partial(_fail_exception, SCRIPT_SHORTNAME)
+CONFIG_OR_EXIT = partial(_config_or_exit, FAIL_EXCEPTION, ERR_BAD_CONFIG)
```
Comment (doctoryes) [inline scripts/bulk_unsuppress_segment_users.py:40]: Forgot about all these! Walk down memory lane...

File: tubular/segment_api.py [hunk 2]
```diff
@@ -140,6 +140,35 @@ def _get_value_from_learner(self, learner, key):
...
+        # eat the TypeError / KeyError since they won't be relevant.
+        except (TypeError, KeyError, requests.exceptions.HTTPError, JSONDecodeError) as exc:
+            LOG.exception(exc)
+            err = u'Error was encountered for for params: {} \n\n Response: {}'.format(
+                params,
+                text_type(resp_json)
+            ).encode('utf-8')
+            LOG.error(err)
+
+            raise Exception(err)
```
Comment (doctoryes) [inline tubular/segment_api.py:170]: There must be much past wisdom buried in this exception handling. 😸

File: tubular/segment_api.py [hunk 3]
```diff
@@ -149,6 +178,51 @@ def delete_learner(self, learner):
...
+        :param key: Key in the learner dict to pull the ID we care about from.
+        :param learners: List of learner dicts to be worked on. We only use the key passed in.
+        :param chunk_size: How many learners should be retired in this batch.
+        :param beginning_idx: Index into learners where this batch should start.
+        """
+        curr_idx = beginning_idx
+        while curr_idx < len(learners):
+            start_idx = curr_idx
+            end_idx = min(start_idx + chunk_size - 1, len(learners) - 1)
+            print("{}, {}, {}".format(start_idx, end_idx, learners))
```
Comment (doctoryes) [inline tubular/segment_api.py]: Is this a debug `print` that was left in?

File: tubular/segment_api.py [hunk 4]
```diff
@@ -149,6 +178,51 @@ def delete_learner(self, learner):
...
+                    "name": "userId",
+                    "values": learner_vals
+                }
+            }
+
+            self._send_regulation_request(params)
+
+            curr_idx += chunk_size
+
     def delete_learners(self, learners, chunk_size, beginning_idx=0):
```
Comment (doctoryes) [inline tubular/segment_api.py]: Nit: Change this func from `delete` to `supress`?

File: tubular/tests/test_segment_api.py [hunk 5]
```diff
@@ -95,23 +95,75 @@ def test_bulk_delete_success(setup_bulk_delete):  # pylint: disable=redefined-ou
...
     mock_post.return_value = FakeErrorResponse()
 
     learner = TEST_SEGMENT_CONFIG['learner']
     with pytest.raises(Exception):
         segment.delete_learners(learner, 1000)
 
     assert mock_post.call_count == 4
-    assert "Error was encountered for learners between start/end indices (0, 0)" in caplog.text
-    assert "{'error': 'Test error message'}" in caplog.text
+    assert "Error was encountered for for params:" in caplog.text
```
Comment (doctoryes) [inline tubular/tests/test_segment_api.py]: `for for params`?

File: tubular/tests/test_segment_api.py [hunk 6]
```diff
@@ -95,23 +95,75 @@ def test_bulk_delete_success(setup_bulk_delete):  # pylint: disable=redefined-ou
...
+    """
+    mock_post, segment = setup_regulation_api
+    mock_post.return_value = FakeErrorResponse()
+
+    learner = TEST_SEGMENT_CONFIG['learner']
+    with pytest.raises(Exception):
+        segment.unsuppress_learners_by_key('original_username', learner, 100)
+
+    assert mock_post.call_count == 4
+    assert "Error was encountered for for params:" in caplog.text
```
Comment (doctoryes) [inline tubular/tests/test_segment_api.py]: `for for params`?

**Review by doctoryes — COMMENTED** [review]
Looks good - only some minor questions...

File: tubular/segment_api.py (hunk 3 above)
Comment (bmedx) [reply tubular/segment_api.py]: Yep, I'll remove

File: tubular/segment_api.py (hunk 4 above)
Comment (bmedx) [reply tubular/segment_api.py]: Technically this is delete AND suppress, but I can make that change.

**Review by doctoryes — APPROVED** [review]

**Pull request merged**